crates/context/config/src/client/starknet.rs

```diff
@@ -0,0 +1,168 @@
+    pub account_id: String,
+    pub access_key: String,
+}
+
+#[derive(Debug)]
+pub struct StarknetConfig<'a> {
+    pub networks: BTreeMap<Cow<'a, str>, NetworkConfig>,
+}
+
+#[derive(Clone, Debug)]
+struct Network {
+    client: Arc<JsonRpcClient<HttpTransport>>,
+    account_id: String,
+    secret_key: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct StarknetTransport<'a> {
+    networks: BTreeMap<Cow<'a, str>, Network>,
+}
+
+impl<'a> StarknetTransport<'a> {
+    #[must_use]
+    pub fn new(config: &StarknetConfig<'a>) -> Self {
+        let mut networks = BTreeMap::new();
+
+        for (network_id, network_config) in &config.networks {
+            let client = JsonRpcClient::new(HttpTransport::new(network_config.rpc_url.clone()));
+            let _ignored = networks.insert(
+                network_id.clone(),
+                Network {
+                    client: client.into(),
+                    account_id: network_config.account_id.clone(),
+                    secret_key: network_config.access_key.clone(),
+                },
+            );
+        }
+        Self { networks }
+    }
+}
+
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum StarknetError {
+    #[error("unknown network `{0}`")]
+    UnknownNetwork(String),
+    #[error("invalid response from RPC while {operation}")]
+    InvalidResponse { operation: ErrorOperation },
+    #[error("invalid contract ID `{0}`")]
+    InvalidContractId(String),
+    #[error("access key does not have permission to call contract `{0}`")]
+    NotPermittedToCallContract(String),
+    #[error(
+        "access key does not have permission to call method `{method}` on contract {contract}"
+    )]
+    NotPermittedToCallMethod { contract: String, method: String },
+    #[error("transaction timed out")]
+    TransactionTimeout,
+    #[error("error while {operation}: {reason}")]
+    Custom {
+        operation: ErrorOperation,
+        reason: String,
+    },
+}
+
+#[derive(Copy, Clone, Debug, Error)]
+#[non_exhaustive]
+pub enum ErrorOperation {
+    #[error("querying contract")]
+    Query,
+    #[error("mutating contract")]
+    Mutate,
+    #[error("fetching account")]
+    FetchAccount,
+}
+
+impl Transport for StarknetTransport<'_> {
+    type Error = StarknetError;
+
+    async fn send(
+        &self,
+        request: TransportRequest<'_>,
+        payload: Vec<u8>,
+    ) -> Result<Vec<u8>, Self::Error> {
+        let Some(network) = self.networks.get(&request.network_id) else {
+            return Err(StarknetError::UnknownNetwork(
+                request.network_id.into_owned(),
+            ));
+        };
+
+        let contract_id = request.contract_id.as_ref();
+
+        match request.operation {
+            Operation::Read { method } => Ok(network
+                .query(contract_id, &method, payload)
+                .await?
+                .into_iter()
+                .flat_map(|felt| felt.to_bytes_be().to_vec())
+                .collect::<Vec<u8>>()),
```

Comment:
why not return `Vec<u8>` from `Network::query` to keep this location neat

Reply:
Updated